File: pattern_matching.py
# // Time Complexity :O(n) where n is the length of the string
# // Space Complexity :O(1) for both hashmap and hashset
# // Did this code successfully run on Leetcode :yes
# // Any problem you faced while coding this :yes


# // Your code here along with comments explaining your approach




class Solution:
    def wordPattern(self, pattern: str, s: str) -> bool:
        dict={}
        s=s.split(" ")
        print(s)
        hashset=[False]*100
        if(len(s))!=(len(pattern)):
            return False
        for i in range(len(pattern)):
            print(s[i])
            if s[i] not in dict.keys():
                print(ord(pattern[i])-ord('a'))
                if hashset[ord(pattern[i])-ord('a')]==True:
                    return False
                else:
                    dict[s[i]]=pattern[i]
                    hashset[ord(pattern[i])-ord('a')]=True
                    print(dict)
            else:
                if dict[s[i]]!=pattern[i]:
                    return False
        return True
            
        
        
            
            
            
        

File: pattern_matching.java
/******************** Pattern Match ****************************/
// Time Complexity : O(n)
// Space Complexity : O(n)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this :

// Your code here along with comments explaining your approach
class Solution {

  public boolean wordPattern(String pattern, String s) {
    //checking for the base case- if either of the length is 0, we can't check the conditions
    if (pattern.length() == 0 || s.length() == 0) return false;
    //split the s string into a string array
    String[] word = s.split(" ");
    if (pattern.length() != word.length) return false; //length has to be same to be compared
    HashMap<Character, String> map = new HashMap<>();

    for (int i = 0; i < pattern.length(); i++) {
      if (map.containsKey(pattern.charAt(i))) { //if the key is present, we check if it is mapped to the corresponding element in word array
        if (!map.get(pattern.charAt(i)).equals(word[i])) {
          return false;
        }
      } else if (map.containsValue(word[i])) { //if the key is not present but the value is present, then it indicates it is mapped to a different character and that breaks the condition
        return false;
      } else {
        map.put(pattern.charAt(i), word[i]); //map the corresponding character in pattern with a string in word
      }
    }
    return true;
  }
}
